File: TypeScript/TP3_MET_TS/src/Interface/Mot.ts
import {IterateurMot} from "./IterateurMot";
import {VisiteurMot} from "./VisiteurMot";
import {IterateurMotInductif} from "../Inductif/IterateurMotInductif";
import {IterateurMotiteratif} from "../Iteratif/IterateurMotiteratif";

export default abstract class Mot implements Iterable<string> {

    // Sélecteurs
    casVide(): boolean {
        return false;
    }

    casCaractereMot(): boolean {
        return false;
    }

    casMotMot(): boolean {
        return false;
    }

    caractere(): string {
        throw new Error();
    }

    reste(): Mot {
        throw new Error();
    }

    getMotGauche(): Mot {
        throw new Error("Erreur");
    }

    getMotDroite(): Mot {
        throw new Error("Erreur");
    }

    abstract creerMotVide(): Mot;

    abstract creerCaractereMot(c: string): Mot;

    abstract creerMotMot(droite: Mot): Mot;

    iterateurInductif(): IterateurMot {
        return new IterateurMotInductif(this);
    }

    iterateurIteratif(): IterateurMot {
        return new IterateurMotiteratif(this);
    }

    iterator(): Iterator<string> {
        return this.iterateurIteratif();
    }

    // Visiteur itératif (programmé récursivement puis itérativement)
    accueilRecursif<T>(v: VisiteurMot<T>): T {
        if (this.casVide()) {
            return v.casVide();
        }
        return v.casCaractereMot(this.caractere(), this.reste().accueilRecursif(v));
    }

    accueil<T>(v: VisiteurMot<T>): T {
        let r: T = v.casVide();
        for (let x in this) {
            r = v.casCaractereMot(x, r);
        }
        return r;
    }

    // Visiteur récursif primitif analogue du filtrage par cas (pattern matching),
    // programmé récursivement
    filtrageRécursif<T>(casVide: () => number, casCaractereMot: (c: string, m: Mot) => (t: T) => T) {
        if (this.casVide()) {
            return casVide.apply;
        }
        return casCaractereMot.apply(this.caractere(), this.reste());
    }

    // Visiteur récursif primitif analogue du filtrage par cas (pattern matching),
    // programmé récursivement
    filtrage<T>(casVide: () => T, casCaractereMot: (c: string, m: Mot) => (t: T) => T) {
        let r: T = casVide();
        let arg: Mot = this.creerMotVide();
        let courant: Mot = this;
        while (!courant.casVide()) {
            let e: string = courant.caractere();
            r = casCaractereMot(e, arg)(r);
            arg = arg.creerCaractereMot(e);
            courant = courant.reste();
        }
        return r;
    }
}

File: Java/TP3_MET_JAVA/src/Interface/Mot.java
package Interface;

import Inductif.IterateurMotInductif;
import Inductif.VisiteurMotInductif;
import Iteratif.IterateurMotiteratif;
import Iteratif.VisiteurMotIteratif;

import java.util.Iterator;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.function.Supplier;

public interface Mot extends Iterable<Character> {

    // Sélecteurs
    default boolean casVide() {
        return false;
    }


    default boolean casCaractereMot() {
        return false;
    }

    default boolean casMotMot() {
        return false;
    }

    Character caractere();

    Mot reste();
    /*
    default Character caractere() {
        throw new UnsupportedOperationException();
    }

    default Mot reste() {
        throw new UnsupportedOperationException();
    }*/

    default Mot getMotGauche() {
        throw new UnsupportedOperationException();
    }

    default Mot getMotDroite() {
        throw new UnsupportedOperationException();
    }

    Mot creerMotVide();


    Mot creerCaractereMot(char c);


    Mot creerMotMot(Mot droite);

    default IterateurMot iterateurInductif() {
        return new IterateurMotInductif(this);
    }

    default IterateurMot iterateurIteratif() {
        return new IterateurMotiteratif(this);
    }

    @Override
    default Iterator<Character> iterator() {
        return this.iterateurIteratif();
    }

    // Visiteur recursif
    default <T> T accueilRecursif(VisiteurMotInductif<T> v) {
        if (this.casVide()) {
            return v.casVide();
        }
        return v.casCaractereMot(this.caractere(), this.reste().accueilRecursif(v));
    }

    // Visiteur itératif
    default <T> T accueil(VisiteurMotIteratif<T> v) {
        T r = v.casVide();
        for (Character x : this) {
            r = v.casCaractereMot(x, r);
        }
        return r;
    }

    // Visiteur récursif primitif analogue du filtrage par cas (pattern matching),
    // programmé récursivement
    default <T> T filtrageRécursif(Supplier<T> casVide, BiFunction<Character, Mot, T> casCaractereMot) {
        if (this.casVide()) {
            return casVide.get();
        }
        return casCaractereMot.apply(this.caractere(), this.reste());
    }

    // Visiteur récursif primitif analogue du filtrage par cas (pattern matching),
    // programmé itérativement
    default <T> T filtrage(Supplier<T> casVide, BiFunction<Character, Mot, Function<T, T>> casCaractereMot) {
        T r = casVide.get();
        Mot arg = this.creerMotVide();
        Mot courant = this;
        while (!courant.casVide()) {
            Character e = courant.caractere();
            r = casCaractereMot.apply(e, arg).apply(r);
            arg = arg.creerCaractereMot(e);
            courant = courant.reste();
        }
        return r;
    }


}
